notebook/gateway/handlers.py

```diff
@@ -0,0 +1,393 @@
+# Copyright (c) Jupyter Development Team.
+# Distributed under the terms of the Modified BSD License.
+
+import os
+import json
+import logging
+from socket import gaierror
+
+from ..base.handlers import APIHandler, IPythonHandler
+from ..utils import url_path_join
+
+from tornado import gen, web
+from tornado.concurrent import Future
+from tornado.ioloop import IOLoop
+from tornado.websocket import WebSocketHandler, websocket_connect
+from tornado.httpclient import HTTPRequest
+from tornado.simple_httpclient import HTTPTimeoutError
+from tornado.escape import url_escape, json_decode, utf8
+
+from ipython_genutils.py3compat import cast_unicode
+from jupyter_client.session import Session
+from traitlets.config.configurable import LoggingConfigurable
+
+# Note: Although some of these are available via NotebookApp (command line), we will
+# take the approach of using environment variables to enable separate sets of values
+# for use with the local Notebook server (via command-line) and remote Gateway server
+# (via environment variables).
+
+GATEWAY_HEADERS = json.loads(os.getenv('GATEWAY_HEADERS', '{}'))
+GATEWAY_HEADERS.update({
+    'Authorization': 'token {}'.format(os.getenv('GATEWAY_AUTH_TOKEN', ''))
+})
+VALIDATE_GATEWAY_CERT = os.getenv('VALIDATE_GATEWAY_CERT') not in ['no', 'false']
+
+GATEWAY_CLIENT_KEY = os.getenv('GATEWAY_CLIENT_KEY')
```

Comment:
I'd like to avoid any environment-only configuration in this repo. All configuration should be done through traitlets. Maybe this warrants a `KernelGateway` configurable class to put these on? The defaults can still come from these environment variables if need be.

Reply:
I see - that makes sense.  Is that how parameters like `--<className>.<parameter>` work?  Some of these are referenced in handlers and managers.  Would this be accomplished by creating a Configurable subclass (e.g., `Gateway`) which would then be referenced via multiple inheritance in the handler/manager classes?

Reply:
Exactly right. It would look a bit like:


```python
from traitlets import Unicode, default
from traitlets.config import LoggingConfigurable


class Gateway(LoggingConfigurable):
    client_key = Unicode(
        config=True,
        help="""gateway client key for...""",
   )
   @default('client_key')
    def _default_client_key(self):
        return os.getenv('JUPYTER_GATEWAY_CLIENT_KEY', '')
```

and then somewhere in NotebookApp:

```python
self.gateway = Gateway(parent=self)
```

> which would then be referenced via multiple inheritance in the handler/manager classes?

I'd use a "has a" relationship, because multiple inheritance of Configurables and Handlers doesn't seem to play very nicely together. So the handlers would access `self.gateway.client_key` etc.

Reply:
Beautiful - thank you for the example.  I've moved the PR into `WIP` until I've had a chance to implement these changes.